### PR TITLE
tcp support

### PR DESCRIPTION
**Pull request by rlcalmeida**

### CODE REVIEW

File: confirm.c [hunk 1]
```diff
@@ -395,38 +398,102 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
...
+	int probe_type = 0; // probe_type can hold 1 (icmp) or 2 (tcp)
 
 	struct sockaddr_in6 dst;
 	dst.sin6_family = AF_INET6;
-	if(pkt->icmpv6->icmp_type == ICMP6_ECHOREPLY) {
+
+	if(pkt->ipv6->ip_nh==IPPROTO_TCP) {
+		// SYN-ACK
+		// TODO (rafael): draft
+		if(pkt->tcp->th_flags != 0x012) return NULL;
```
Comment (cunha) [inline confirm.c]: create a constant to store what 0x012 means?

File: confirm.c [hunk 2]
```diff
@@ -395,38 +398,102 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
...
+		probe_type = 2;
```
Comment (cunha) [inline confirm.c]: what is 2? use defines from the top of the file.

File: confirm.c [hunk 3]
```diff
@@ -395,38 +398,102 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
...
-		struct libnet_icmpv6_hdr *ricmp;
-		rip = (struct libnet_ipv6_hdr *)(pkt->payload);
-		ricmp = (struct libnet_icmpv6_hdr *)(pkt->payload + LIBNET_IPV6_H);
-		memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
-		uint32_t flags = *(uint32_t *)(rip->ip_flags);
-		traffic_class = (flags & 0x0FF00000) >> 20;
-		flow_label = (flags & 0x000FFFFF);
-		icmpid = ntohs(ricmp->id);
-		data = ntohs(ricmp->seq);
+		uint32_t tcp_sequence_number = ntohl(pkt->tcp->th_ack) - 1;
```
Comment (cunha) [inline confirm.c]: add a comment to say that the acknowledgement's sequence number is 1 larger than the original?

File: confirm.c [hunk 4]
```diff
@@ -395,38 +398,102 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
...
+		icmpid = (uint16_t) ((tcp_sequence_number >> 16) & 0x0000FFFF);
+		data = (uint16_t) (tcp_sequence_number & 0x0000FFFF);
+		logd(LOG_DEBUG, "%s TCP SYN-ACK received\n", __func__);
```
Comment (cunha) [inline confirm.c]: move this to line 416? it can replace the comment.

File: confirm.c [hunk 5]
```diff
@@ -395,38 +398,102 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
...
-		rip = (struct libnet_ipv6_hdr *)(pkt->payload);
-		ricmp = (struct libnet_icmpv6_hdr *)(pkt->payload + LIBNET_IPV6_H);
-		memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
-		uint32_t flags = *(uint32_t *)(rip->ip_flags);
-		traffic_class = (flags & 0x0FF00000) >> 20;
-		flow_label = (flags & 0x000FFFFF);
-		icmpid = ntohs(ricmp->id);
-		data = ntohs(ricmp->seq);
+		uint32_t tcp_sequence_number = ntohl(pkt->tcp->th_ack) - 1;
+		icmpid = (uint16_t) ((tcp_sequence_number >> 16) & 0x0000FFFF);
```
Comment (cunha) [inline confirm.c]: this variable has to change name, as this is a TCP probe (and a TCP reply). an idea would be to rename `data` and `icmpid` to denote the information that they store.

File: confirm.c [hunk 6]
```diff
@@ -395,38 +398,102 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
...
+		data = (uint16_t) (tcp_sequence_number & 0x0000FFFF);
+		logd(LOG_DEBUG, "%s TCP SYN-ACK received\n", __func__);
+	}
+	else if(pkt->ipv6->ip_nh==IPPROTO_ICMP6) {
+		if(pkt->icmpv6->icmp_type == ICMP6_ECHOREPLY) {
+			probe_type = 1;
+			// Echo reply
```
Comment (cunha) [inline confirm.c]: no need for this comment (line 427 already says it is an echo reply

File: confirm.c [hunk 7]
```diff
@@ -395,38 +398,102 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
...
-		icmpid = ntohs(ricmp->id);
-		data = ntohs(ricmp->seq);
+		uint32_t tcp_sequence_number = ntohl(pkt->tcp->th_ack) - 1;
+		icmpid = (uint16_t) ((tcp_sequence_number >> 16) & 0x0000FFFF);
+		data = (uint16_t) (tcp_sequence_number & 0x0000FFFF);
+		logd(LOG_DEBUG, "%s TCP SYN-ACK received\n", __func__);
+	}
+	else if(pkt->ipv6->ip_nh==IPPROTO_ICMP6) {
+		if(pkt->icmpv6->icmp_type == ICMP6_ECHOREPLY) {
+			probe_type = 1;
```
Comment (cunha) [inline confirm.c]: use defines

File: confirm.c [hunk 8]
```diff
@@ -395,38 +398,102 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
...
+		}
+		else if((pkt->icmpv6->icmp_type == ICMP6_DST_UNREACH) &&
+			(pkt->icmpv6->icmp_code == ICMP6_DST_UNREACH_NOPORT)){
+			// Port unreachable
+			// TODO (rafael): draft (to be tested)
+			probe_type = 2;
+			struct libnet_ipv6_hdr *rip = (struct libnet_ipv6_hdr *)(pkt->payload);
+			struct libnet_tcp_hdr *rtcp = (struct libnet_tcp_hdr *)
+					(pkt->payload + LIBNET_IPV6_H);
+			memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
```
Comment (cunha) [inline confirm.c]: don't we need an `if` here to check whether `rip` is ICMP or TCP? i understand that only TCP can elicit `UNREACH_NOPORT`, but ICMP may elicit `UNREACH` responses.

File: confirm.c [hunk 9]
```diff
@@ -436,9 +503,23 @@ static struct confirm_query * confirm_pkt_parse6(const struct packet *pkt)/*{{{*
...
 
 	struct sockaddr_storage *ptr = (struct sockaddr_storage *)&dst;
-	struct confirm_query *q = confirm_query_create6(ptr, ttl,
-			traffic_class, flow_label,
-			icmpid, flowid, NULL);
+	struct confirm_query *q;
+
+	logd(LOG_DEBUG, "received probe %d ttl %d flowid %d fixrev %d\n", probe_type, ttl, flowid, fixrev);
+
+	if(probe_type==1){
```
Comment (cunha) [inline confirm.c]: defines

File: confirm.c [hunk 10]
```diff
@@ -597,47 +678,58 @@ static void event_run_sendpacket(struct confirm *conf, struct event *ev)
...
-					query->padding);
-		}
-		else {
-			logd(LOG_FATAL, "%s %d: fixrev for IPv6 not impl\n",
-					__FILE__, __LINE__);
-			pkt = NULL;
+	if(query->type==QUERY_TYPE_TCP){
+		struct sockaddr_in6 *dst = (struct sockaddr_in6 *) &(query->dst);
+		struct libnet_in6_addr ipv6_dst;
+		memcpy(&ipv6_dst, &(dst->sin6_addr), sizeof(ipv6_dst));
```
Comment (cunha) [inline confirm.c]: do you need `dst`? can't you copy straight from `query->dst`?

File: confirm.c [hunk 11]
```diff
@@ -763,9 +855,35 @@ confirm_query_create4(const struct sockaddr_storage *dst, uint8_t ttl,
...
+	query->src_port = src_port;
+	query->dst_port = dst_port;
+	query->type = QUERY_TYPE_TCP;
+	return query;
+}
+
+struct confirm_query *
+confirm_query_create6_icmp(const struct sockaddr_storage *dst, uint8_t ttl,
+		uint8_t traffic_class, uint32_t flow_label, uint16_t icmpid,
+		uint8_t flowid, confirm_query_cb cb)
```
Comment (cunha) [inline confirm.c]: why don't we have `icmpseq` here?

File: confirm.c [hunk 12]
```diff
@@ -763,9 +855,35 @@ confirm_query_create4(const struct sockaddr_storage *dst, uint8_t ttl,
...
+	query->icmpid = icmpid;
+	query->type = QUERY_TYPE_ICMP;
+	return query;
+}
+
+struct confirm_query *
 confirm_query_create6(const struct sockaddr_storage *dst, uint8_t ttl,
-		uint8_t traffic_class, uint32_t flow_label,
-		uint16_t icmpid, uint8_t flowid,
+		uint8_t traffic_class, uint32_t flow_label, uint8_t flowid,
```
Comment (cunha) [inline confirm.c]: if it is not actively used, I would try to remove this function and call the `_icmp` and `_tcp` versions. (i don't remember what else is in the code though)

File: packet.c [hunk 13]
```diff
@@ -64,49 +64,59 @@ static void packet4_fill(struct packet *pkt, size_t ipoffset)/*{{{*/
 static void packet6_fill(struct packet *pkt, size_t ipoffset)/*{{{*/
 {
 	pkt->ipv6 = (struct libnet_ipv6_hdr *)(pkt->buf + ipoffset);
-	pkt->icmpv6 = (struct libnet_icmpv6_hdr *)
+	
+	if(pkt->ipv6->ip_nh==IPPROTO_TCP){
```
Comment (cunha) [inline packet.c]: can we put this whole if inside the switch below and make the code simpler?

File: sender6.c [hunk 14]
```diff
@@ -96,7 +102,7 @@ struct packet * sender6_send_icmp(struct sender6 *s, /* {{{ */
 
 	if(libnet_write(s->ln) < 0) goto out;
 
-	struct packet *pkt = sender6_make_packet(s);
+	struct packet *pkt = sender6_make_packet(s, 0);
```
Comment (cunha) [inline sender6.c]: what is 0? use defines

File: sender6.c [hunk 15]
```diff
@@ -143,3 +160,59 @@ static struct packet * sender6_make_packet(struct sender6 *s)/*{{{*/
...
+	return retval;
+}
+
+// TODO (rafael): draft
+struct packet * sender6_send_tcp(struct sender6 *s, struct libnet_in6_addr dst,
+		uint8_t ttl, uint8_t traffic_class, uint32_t flow_label, uint8_t flowid,
+		uint16_t sp, uint16_t dp)
+{
+
+	uint32_t seq = (0 << 16) | sender6_compute_tcp_sequence(ttl, flowid, 0);
```
Comment (cunha) [inline sender6.c]: `(0 << 16)` does not do anything, it is safe to remove

File: sender6.c [hunk 16]
```diff
@@ -143,3 +160,59 @@ static struct packet * sender6_make_packet(struct sender6 *s)/*{{{*/
...
+
+	uint32_t ack = 0;
+	uint8_t control = 0x002;
```
Comment (cunha) [inline sender6.c]: defines for `0x002`?

File: sender6.c [hunk 17]
```diff
@@ -143,3 +160,59 @@ static struct packet * sender6_make_packet(struct sender6 *s)/*{{{*/
...
+	uint16_t win = 5760;
+	uint16_t sum = 0;
+	uint16_t urg = 0;
+	uint16_t h_len = LIBNET_TCP_H + 20;
```
Comment (cunha) [inline sender6.c]: what is the 20?

File: sender6.c [hunk 18]
```diff
@@ -143,3 +160,59 @@ static struct packet * sender6_make_packet(struct sender6 *s)/*{{{*/
...
+	const uint8_t *payload = NULL;
+	uint32_t payload_s = 0;
+
+	// TODO (rafael): draft
+	s->opttag = libnet_build_tcp_options((uint8_t*)"\003\003\002\004\002\001\002\004\001\011\000\000\000\000\000\000\000\000\000\000", 20, s->ln, s->opttag);
```
Comment (cunha) [inline sender6.c]: holy cow what is this? =D

File: sender6.c [hunk 19]
```diff
@@ -143,3 +160,59 @@ static struct packet * sender6_make_packet(struct sender6 *s)/*{{{*/
...
+
+	size_t sz = LIBNET_TCP_H + 20;
+	s->iptag = libnet_build_ipv6(traffic_class, flow_label, sz, 6, ttl, s->ip,
+		dst, NULL, 0, s->ln, s->iptag);
+
+	if(s->iptag == -1) goto out;
+
+	if(libnet_write(s->ln) < 0) goto out;
+
+	struct packet *pkt = sender6_make_packet(s, 1);
```
Comment (cunha) [inline sender6.c]: 1? (these constants make it hard to follow what is going on)

File: sender6.h [hunk 20]
```diff
@@ -23,4 +23,8 @@ struct packet * sender6_send_icmp_fixrev(struct sender6 *sender,
 		size_t padding);
 */
 
+struct packet * sender6_send_tcp(struct sender6 *s, struct libnet_in6_addr dst,
```
Comment (cunha) [inline sender6.h]: like the `send_icmp` function, this function should receive values for the TCP packer directly. in particular, it should not receive `flowid`, but should receive the sequence number. this is so the function is generic and can be used in other modules (you should generate the value for the sequence number (encoding flow id, ttl, and sequence) outside this module).

File: testlib.c [hunk 21]
```diff
@@ -32,20 +32,35 @@ void querycb(struct confirm_query *q)/*{{{*/
...
+	}
+
+	if((ipversion!=4) && (ipversion!=6)){
+		printf("unknown ip version\n");
+		exit(EXIT_FAILURE);
 	}
-	else{
-		ipversion = 4;
+
+	if((probe_type==2) && (ipversion==4)){
```
Comment (cunha) [inline testlib.c]: `probe_type` should be compared to the defines. another idea is to use the protocol number, as defined in the standards.

File: confirm.c (hunk 11 above)
Comment (rawfael) [inline confirm.c]: the previous version of netlib already worked that way. I also think we should have icmp_seq here, we should have seq_number in confirm_query_create6_tcp too. Since we are packing data in these fields, they dont appear here. I think we should remove data_pack and data_unpack from netlib, since these functions are not generic (they are specific to our application paris6). That way we can add these fields to confirm_query_create6_icmp and confirm_query_create6_tcp. Can I do that?

Comment (cunha) [inline confirm.c]: i like this idea. i am not sure whether it is better to remove
data_pack/unpack from the library entirely (as we would need to update
dtrack and remaproute to do the packing and unpacking of data). a good
trade-off could be to move it inside confirm.c (which is the module that
needs to do the mapping of probe-to-response)

File: confirm.c (hunk 12 above)
Comment (rawfael) [inline confirm.c]: I tried to put the common things to tcp and icmp in this function. I think it should be a static function, right?

Comment (cunha) [inline confirm.c]: It might be better to have two... a confirm_query_create_tcp6 and _icmp6. I
understand this is not super clean, but it is the way netlib does it, so at
least people will need to get familiar with one interface only.

File: confirm.c [hunk 22]
```diff
@@ -395,50 +398,123 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
...
 	struct sockaddr_storage *ptr = (struct sockaddr_storage *)&dst;
-	struct confirm_query *q = confirm_query_create6(ptr, ttl,
-			traffic_class, flow_label,
-			icmpid, flowid, NULL);
+	struct confirm_query *q;
+
+	logd(LOG_DEBUG, "received probe %d ttl %d flowid %d fixrev %d\n", probe_type, ttl, flowid, fixrev);
+
+	if(probe_type==PROBE_TYPE_ICMP){
+		q = confirm_query_create6_icmp(ptr, ttl, traffic_class, flow_label,
```
Comment (cunha) [inline confirm.c]: one problem with these `if`s is that we do not fill in `traffic_class` and `flow_label` for TCP SYN-ACK and ICMP ECHOREPLY messages.

Given we only use `traffic_class` and `flow_label` to identify load balancers, can't we compute their values from the flow-ID stored in `packed_data`? The vanilla version of Paris traceroute does this. If we used the flow-ID to fill in `traffic_class` and `flow_label`, we would (1) always have this information and (2) be able to check whether the original packet was mangled in any way.

Does this make sense? Do you think we can implement it?

Comment (cunha) [inline confirm.c]: This is less important for SYN-ACK replies, as they should pre very rare; but it would be a common for ICMP echo replies.

File: confirm.c [hunk 23]
```diff
@@ -597,47 +673,65 @@ static void event_run_sendpacket(struct confirm *conf, struct event *ev)
...
-					revsum, data,
-					query->padding);
-		}
-		else {
-			logd(LOG_FATAL, "%s %d: fixrev for IPv6 not impl\n",
-					__FILE__, __LINE__);
-			pkt = NULL;
+		uint32_t ack_number = 0;
+		uint8_t control_flags = TH_SYN;
+		uint16_t windows_size = 5760;
```
Comment (cunha) [inline confirm.c]: where did 5760 come from?

File: confirm.c [hunk 24]
```diff
@@ -597,47 +673,65 @@ static void event_run_sendpacket(struct confirm *conf, struct event *ev)
...
-					query->ipid,
-					id2checksum[query->flowid],
-					revsum, data,
-					query->padding);
-		}
-		else {
-			logd(LOG_FATAL, "%s %d: fixrev for IPv6 not impl\n",
-					__FILE__, __LINE__);
-			pkt = NULL;
+		uint32_t ack_number = 0;
```
Comment (cunha) [inline confirm.c]: shouldn't this include data?

File: confirm.c [hunk 25]
```diff
@@ -763,15 +857,19 @@ confirm_query_create4(const struct sockaddr_storage *dst, uint8_t ttl,
...
+confirm_query_create6_tcp(const struct sockaddr_storage *dst, uint8_t ttl,
+		uint8_t traffic_class, uint32_t flow_label, uint8_t flowid,
+		uint16_t src_port, uint16_t dst_port, confirm_query_cb cb)
 {
-	struct confirm_query *query;
+	struct confirm_query *query = malloc(sizeof(struct confirm_query));
 
-	query = malloc(sizeof(struct confirm_query));
-	if(!query) logea(__FILE__, __LINE__, NULL);
+	if(!query){
```
Comment (cunha) [inline confirm.c]: my argument is that error checking is boring and happens all the time; so I try to reduce the amount of space it takes int he code.

File: confirm.c [hunk 26]
```diff
@@ -780,19 +878,64 @@ confirm_query_create6(const struct sockaddr_storage *dst, uint8_t ttl,
...
 		logd(LOG_WARN, "%s,%d: flowid > 127!\n", __FILE__, __LINE__);
 	}
+
+	memcpy(&(query->dst), dst, sizeof(query->dst));
+	memset(&(query->ip), UINT8_MAX, sizeof(query->ip));
+	query->ip.ss_family = dst->ss_family;
+	query->ttl = ttl;
+	query->traffic_class = traffic_class;
+	query->flow_label = flow_label;
+	query->src_port = 0;
```
Comment (cunha) [inline confirm.c]: does it make sense to use a `union` to avoid having to set port numbers on ICMP packets and IDs on TCP packets?

File: confirm.h [hunk 27]
```diff
@@ -27,6 +28,9 @@ struct confirm_query {/*{{{*/
 		};
 	};
 
+	uint16_t src_port;
```
Comment (cunha) [inline confirm.h]: here is where I think a `union` could help.

File: confirm.c (hunk 22 above)
Comment (rawfael) [inline confirm.c]: Do you mean to get the traffic_class and flowlabel from the probe that was sent? If so, these values are not being used in order to match the responses to probes (in query_cmp), perhaps we could just remove them.

File: confirm.c (hunk 24 above)
Comment (rawfael) [inline confirm.c]: Data are being sent in sequence number. I will create a seq_number variable for this become clear then.

File: confirm.c (hunk 26 above)
Comment (rawfael) [inline confirm.c]: Yes. In query_cmp we compare icmpid, can I remove that? Are address, ttl and flowid sufficient to match (I think so)?

Comment (cunha) [inline confirm.c]: this seems enough for now. in the future, we might want to pack the try number (see `query->ntries`) in the probe (currently, when we retry a probe and send it again because of timeouts, we cannot identify which try was answered)

File: confirm.c (hunk 24 above)
Comment (cunha) [inline confirm.c]: my bad, mixed up sequence number and acknowledgement.

File: confirm.c (hunk 22 above)
Comment (cunha) [inline confirm.c]: not sure whether we should remove them (depends on whether this information is useful in the rest of the code). my comment though, was about creating a mapping of `flowid` to both `traffic_class` and `flowlabel`. for an example of this for IPv4 and ICMP, see here:

https://github.com/cunha/netlib/blob/master/confirm.c#L30

**Comment by taiar** [comment]
Great code review! :+1:
